### PR TITLE
Real time logging and other improvements

### PR DESCRIPTION
**Pull request by sgrebnov**
1. Added cordova-common usage to improve logging, simplify working w/ plugins, config, etc
2. Implemented paramedic-plugin which provides the following functionality
 - Web socket connectivity to paramedic server
 - Console calls proxy
 -  Injects custom Jasmine reporter which can proxy all calls to server side (enabled real time spec log  in Paramedic server using standard Jasmine reporters – console, junit, other Jasmine reporters)
Demo: https://onedrive.live.com/redir?resid=F25E4AD02C3EAF0D!12364&authkey=!ALyI21Zy_H6yA_I&ithint=video%2cmov

3. Re-wrote local server to use socket.io
 - ‘Disconnected’ event is used to detect when app is crashed or there is connectivity issue. So we don’t need to wait for some time to detect that
 - Real time spec log helps detecting what test caused app crash/etc as we have real time spec logging now
4. Support of `.paramedic.config.js` (paramedic tests configuration file), several test targets, ability to specify test device and other arguments

5. Documentation improvements

Verified that original functionality continues to work, below you can find example output for travis auto-tests:
https://travis-ci.org/tommsawyer/cordova-paramedic/builds/113155397

image

### CODE REVIEW

**Comment by sgrebnov** [comment]
@tommsawyer, could you please sign Apache ICLA and report back once this is done:
1. Complete http://www.apache.org/licenses/icla.pdf
2. Submit to secretary@apache.org

**Comment by nikhilkh** [comment]
@sarangan12, @dblotsky to take a look too

**Comment by purplecabbage** [comment]
I would rather not introduce the socket.io dependency as it has issues on some windows machines. VS seems to be a requirement to compile on install.

**Comment by sgrebnov** [comment]
@purplecabbage You are right about VS requirement, didn't realize this. But I really like that we can detect app crash w/o waiting for timeout and how socketio simplifies development. Does someone know any good alternative? Also, taking into account windows host most likely will be used for cordova-windows tests this should not be a big problem.

File: lib/Tunnel.js [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+    var self = this;
+    //TODO: use localtunnel module instead of shell
+    return Q.Promise(function(resolve, reject) {
+        exec(path.resolve(__dirname, '../node_modules/.bin/lt') + ' --port ' + self.port, null, function(output) {
+            resolve(output.split(' ')[3]);
+        });
+    });
+};
+
+module.exports = Tunnel;
```
Comment (dblotsky) [inline lib/Tunnel.js]: Newline at the end of the file.

**Comment by dblotsky** [comment]
Some general design comments:
- this code looks overly complex; there's a 3-fold increase in code size
- there are too few comments; there are 10 new classes and it's hard to get a clear picture of what they all do
- the API and class hierarchy is unclear; there are at least three classes with a method for installing plugins
- what's the role of the paramedic plugin? does it do similar things to the test framework plugin?

File: sample-config/.paramedic-core-plugins.config.js [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+             "args": "--device"
+        },
+        {    // Windows 8.1 Desktop(anycpu)
+            "platform": "windows@https://github.com/apache/cordova-windows.git",
+            "action": "run"
+        },
+        {   // Windows 10 Desktop(x64)
+            "platform": "windows@https://github.com/apache/cordova-windows.git",
+            "action": "run",
+            "args": "--archs=x64 -- --appx=uap"
```
Comment (nikhilkh) [inline sample-config/.paramedic-core-plugins.config.js]: @sarangan12 Looks like paramedic already supports passing arbitrary args to `cordova run`

**Comment by sgrebnov** [comment]
Hey @dblotsky thx for review, answering your notes

1. The code size has been increased as there is now more functionality here, for example support for jasmine reporters (junit for jenkins), ability to test more than one platform and aggregate results. I found out this very useful  - I currently run tests for Android on two devices (5.x, 6.x) and different Windows versions (10, 8.1, 8.1-phone) so the app is created and the plugins are installed only once which is much faster

2.  Paramedic plugin is different, it adds support for real-time jasmine reporters on nodejs side (works on top of test-framework to proxy jasmine calls to server side)

**Comment by sgrebnov** [comment]
Are there any blockers to complete this PR (api surface/functionality, msbuild requirement on windows)? @purplecabbage Jesse, I primary looking on your opinion as original author.
If there are no blockers here, I'll improve code documentation and do re-base.

**Comment by dblotsky** [comment]
@sgrebnov is the real-time reporting worth the increase in complexity? I personally prefer simpler design, but it's yours and @purplecabbage's call. I'm just interested in the reasoning.

**Comment by nikhilkh** [comment]
One of the advantages of real-time logging is that a native crash in iOS or Android can let me you know the last spec that ran before the crash and helps with attribution on what spec caused the crash.

**Comment by dblotsky** [comment]
@nikhilkh regular native logs will contain all of this information as well though, won't they? So far they've been pretty good at helping us debug crashes.

**Comment by sgrebnov** [comment]
@dblotsky could you please elaborate on `regular native logs` in paramedic (for example when I use paramedic with travis ci)

**Comment by nikhilkh** [comment]
There are lots of scenarios where we do not have native logs currently:
- Windows Phone 10 emulator or device
- Windows Phone 8.1 emulator or device
- iOS device

There are probably others - if it's easily available in console it's better than hunting through native logs anyway.

**Comment by dblotsky** [comment]
@sgrebnov I mean ADB logs on Android and system logs on iOS. @nikhilkh is right though: we don't have easy access to logs on Windows yet (although Windows doesn't work on Travis anyway). Also from what I understand we're already working on adding logging to paramedic. Regardless though, the complexity argument is purely a matter of my own opinion. If you see value in having real-time reporting and you don't mind the slight increase in maintainability, it's all good.

**Comment by nikhilkh** [comment]
This is a huge set of changes and we have more changes in queue as we look to run paramedic as our solution for running PR builds. What are the next steps here? 

As for socketio requiring VS on Windows - I don't see any native node module compile which generally causes that.

```
λ npm install https://github.com/sgrebnov/cordova-paramedic-old
cordova-paramedic@0.5.0 node_modules\cordova-paramedic
├── q@1.4.1
├── ansi@0.3.1
├── minimist@1.1.3
├── tmp@0.0.25
├── shelljs@0.3.0
├── jasmine-reporters@2.1.1 (mkdirp@0.3.5)
├── jasmine-spec-reporter@2.4.0 (colors@1.1.2)
├── socket.io@1.4.5 (has-binary@0.1.7, debug@2.2.0, socket.io-parser@2.2.6, socket.io-adapter@0.4.0, engine.io@1.6.8, socket.io-client@1.4.5)
├── localtunnel@1.5.1 (debug@0.7.4, optimist@0.3.4, request@2.11.4)
├── request@2.69.0 (aws-sign2@0.6.0, forever-agent@0.6.1, tunnel-agent@0.4.2, caseless@0.11.0, is-typedarray@1.0.0, oauth-sign@0.8.1, stringstream@0.0.5, isstream@0.1.2, json-stringify-safe@5.0.1, extend@3.0.0, tough-cookie@2.2.2, node-uuid@1.4.7, qs@6.0.2, combined-stream@1.0.5, mime-types@2.1.10, form-data@1.0.0-rc4, aws4@1.3.2, hawk@3.1.3, bl@1.0.3, http-signature@1.1.1, har-validator@2.0.6)
└── cordova-common@1.1.0 (unorm@1.4.1, cordova-registry-mapper@1.1.15, underscore@1.8.3, semver@5.1.0, osenv@0.1.3, bplist-parser@0.1.1, shelljs@0.5.3, glob@5.0.15, elementtree@0.1.6, plist@1.2.0)
```

@purplecabbage Can you please review? It will be great to close on this.

**Comment by purplecabbage** [comment]
Yeah, I echo @dblotsky 's concern here. The fact that the pull request is 2X bigger than the original code is a bit troubling. Unfortunately I haven't had time to discuss alternatives, and now I am in Hawaii (poor me)
I do consider the socketio issue a blocker, we have had issues with it in the PhoneGap-cli since we added it. @dblotsky mentions Travis and Windows; we use appveyor for Windows, so the same rules apply. 
Re: realtime logging, logs are currently live, but paramedic does not dump anything until it receives a result or timeout. We can change this without socketio.

**Comment by purplecabbage** [comment]
Here is the socketio/node-gyp install time issue. 
https://github.com/nodejs/node-v0.x-archive/issues/8005

**Comment by nikhilkh** [comment]
Looks like the native module dependency is not there anymore: 
https://github.com/websockets/ws/commit/49b11093e9a009e5305dcde7003d3a896b2811dc

socket.io used to depend on ws which depended on something that require native compilation. This is consistent with my current install experience. Can anyone else try this on Windows and see if they get a native node module compile experience requiring VS.

**Comment by dblotsky** [comment]
+1 to Jesse's comment about logging. I'm all for not using socket.io if paramedic already does live logging.

**Comment by sgrebnov** [comment]
I'll test socket.io installation on clean machine as I think native dependencies / msbuild requirement is no more actual.

**Comment by sarangan12** [comment]
@sgrebnov Is there a reason to support multiple targets in the same config file? Is that scenario really helpful for developers? Wouldn't it be better to leave it at just one target per config file. (This would remove the requirement of removing the platform for every target after run)

**Comment by sgrebnov** [comment]
I've reviewed `socket.io` dependencies and also tested on clean machine and confirm there are NO native dependencies here anymore.

**Comment by sgrebnov** [comment]
@sarangan12 I currently run auto tests for cordova core plus a few custom plugins for several platforms and this helps a lot.

**Comment by dblotsky** [comment]
@sgrebnov can't you just invoke it with different parameters? CLIs usually support a single set of arguments, and if you want different arguments per call you just pass them to separate invocations. This makes the design and code a lot simpler.

**Comment by vladimir-kotikov** [comment]
My 2 cents about concerns on this PR:

1. Socket.io - I confirm this is no issue anymore since websockets/ws@49b1109. I don't see any `node-gyp` invocations during socket.io installation on a clean machine. Also an advantage of `socket.io` is that it's well-known, popular, with proven reliability.
2. Multiple targets per config - +1 to @sarangan12 and @dblotsky here, but IMO this shouldn't be a blocker. This doesn't break anything and we just could remove this feature later.
3. Realtime logging/messaging - i believe this is very useful feature as it brings the ability to extend current implementatio in a way we want to. An example is @sarangan12's commit  https://github.com/apache/cordova-paramedic/pull/1/commits/82573075ea9851d73006955bb5d1de5853ebc4d1 adding JUnit reporter.

So i vote for merging this and address any concerns/feedback after.

File: main.js [hunk 3]
```diff
@@ -1,43 +1,51 @@
...
+    "`MSECS` : (optional) time in millisecs to wait for tests to pass|fail \n" +
+              "\t(defaults to 10 minutes) \n" +
+    "`PORTNUM` : (optional) ports to find available and use for posting results from emulator back to paramedic server(default is from 8008 to 8009)\n" +
+    "--justbuild : (optional) just builds the project, without running the tests \n" +
     "--browserify : (optional) plugins are browserified into cordova.js \n" +
     "--verbose : (optional) verbose mode. Display more information output\n" +
-    "--platformPath : (optional) path to install platform from, git or local file uri";
+    "--useTunnel : use tunneling instead of local address. default is false\n" +
+    "--reportSavePath: (optional) path to save Junit results file\n" +
+    "--cleanUpAfterRun: (optional) cleans up the application after the run."
```
Comment (dblotsky) [inline main.js]: Please add the docs for the `--config` flag.

File: lib/ParamedicConfig.js [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+    });
+};
+
+ParamedicConfig.parseFromFile = function (paramedicConfigPath) {
+    return new ParamedicConfig(require(paramedicConfigPath));
+};
+
+ParamedicConfig.prototype.useTunnel = function () {
+    return this._config.useTunnel;
+
```
Comment (dblotsky) [inline lib/ParamedicConfig.js]: Extra newline.

File: lib/reporters/ConsoleReporter.js [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+      for (var i = 0; i < result.failedExpectations.length; i++) {
+        printNewline();
+        print(colored('red', 'An error was thrown in an afterAll'));
+        printNewline();
+        print(colored('red', 'AfterAll ' + result.failedExpectations[i].message));
+
+      }
+      printNewline();
+    }
+  }
```
Comment (dblotsky) [inline lib/reporters/ConsoleReporter.js]: Please fix the indentation in this file.

File: lib/specReporters.js [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+var jasmineReporters    = require('jasmine-reporters');
+
+// default reporter which is used to determine whether tests pass or not
+var paramedicReporter = null;
+
+// all reporters including additional reporters to trace results to console, etc
+var reporters = [];
+
+// specifies path to save Junit results file
+var reportSavePath = null;
```
Comment (dblotsky) [inline lib/specReporters.js]: Please make these variables local instead of global.

File: lib/utils.js [hunk 7]
```diff
@@ -0,0 +1,22 @@
+var shelljs = require('shelljs');
+var verbose = undefined;
```
Comment (dblotsky) [inline lib/utils.js]: Please set this to `null` or `false` instead of `undefined`.

File: sample-config/.paramedic.config.js [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+            "action": "run",
+            "args": "--archs=x64 -- --appx=uap"
+        },
+        // {    // WP 8.1 Device(arm)
+        //     "platform": "windows@https://github.com/apache/cordova-windows.git",
+        //     "action": "run",
+        //     "args": "--archs=arm --device -- --phone"
+        // }
+    ]
+};
```
Comment (dblotsky) [inline sample-config/.paramedic.config.js]: Would these files be better as JSON instead of JS?

**Comment by sgrebnov** [comment]
Taking into account this is dev repo and there is other work pending some changes from this PR I will merge it later today and will help to quickly address all design/api concerns and notes. Let's sync up and discuss api surface change - @vladimir-kotikov  and me will post out thoughts here.

**Pull request merged**